pkg/scheduler/framework/plugins/nodevolumelimits/csi.go

```diff
@@ -134,13 +137,28 @@ func (pl *CSILimits) Filter(ctx context.Context, _ *framework.CycleState, pod *v
 }
 
 func (pl *CSILimits) filterAttachableVolumes(
-	csiNode *storagev1.CSINode, volumes []v1.Volume, namespace string, result map[string]string) error {
+	pod *v1.Pod, csiNode *storagev1.CSINode, volumes []v1.Volume, namespace string, result map[string]string) error {
 	for _, vol := range volumes {
-		// CSI volumes can only be used as persistent volumes
-		if vol.PersistentVolumeClaim == nil {
+		// CSI volumes can only be used through a PVC.
+		pvcName := ""
+		ephemeral := false
+		switch {
+		case vol.PersistentVolumeClaim != nil:
+			pvcName = vol.PersistentVolumeClaim.ClaimName
+		case vol.Ephemeral != nil:
+			if !utilfeature.DefaultFeatureGate.Enabled(features.GenericEphemeralVolume) {
+				return fmt.Errorf(
+					"volume %s is a generic ephemeral volume, but that feature is disabled in kube-scheduler",
```

Comment:
Should an error like this be logged at the scheduler level, or kubelet/api validation? Featuregates are for the whole cluster, so this isn't something that would just be disabled in kube-scheduler

Reply:
You can always have version or feature gate skew. Also, the pod might have been created when the feature was enabled for the entire cluster, then it got disabled and the pod remained in etcd.

It is debatable whether the limits check then should simply skip over the volume or complain. The former is closer to the behavior before the feature was introduced, but the pod is not going to be usable once the feature gets disabled and then a clear error message is better. We decided to do the latter (see commit d2cc70ee2caf628e942ed5d4c49b810401c08acc) so here I am doing the same.

Reply:
> the pod might have been created when the feature was enabled for the entire cluster, then it got disabled and the pod remained in etcd.

this might be worth considering, because I *think* we are generally supposed to be resilient to stored objects in etcd. This could break things if the feature gate is enabled then disabled right?

Reply:
The code is resilient against lingering Pod objects with the disabled volume type, that's why we have this check here. What this code here does is provide a clear error message instead of silently falling through and letting the Pod fail later.

It doesn't break the scheduler, only the Pod, but the Pod is already broken when the feature gets disabled.

Reply:
My comment was about new pods with ephemeral volumes. What I hadn't considered are old, running pods with ephemeral volumes. Those most be counted regardless of the feature gate. Throwing an error would affect all new pods, including those which don't use ephemeral volumes.

I added that distinction and the corresponding unit tests.